resize.js

```diff
@@ -183,17 +198,57 @@ image.encodeAndUpload = function (fileName, fileType, resolutionX, resolutionY,
 
     // Get the image and resize it
     res.writeHead(200, {'Content-Type': supportedFileType(fileType)});
+
     gm(file)
-      .options({imageMagick: true})
-      .autoOrient()
-      .resize(resolutionX, resolutionY)
-      .stream(fileType, function (err, stdout, stderr) {
-        var r = stdout.pipe(res);
-        r.on('finish', function () {
-          // This is to close the result while a background job will continue to process
-          log('info', 'Finished sending a converted image');
-          res.end();
-        });
+      .size(function (err, size) {
+        var originalRatio = size.width / size.height;
+        var newRatio = resolutionX / resolutionY;
+
+        var cropX = 0;
+        var cropY = 0;
+        var cropWidth = size.width;
+        var cropHeight = size.height;
+
+        if (fit === 'crop') {
+          if (originalRatio > newRatio) {
+            var resizeFactor = size.height / resolutionY;
+            cropWidth = size.width / resizeFactor;
+            cropHeight = resolutionY;
+            cropX = (cropWidth - resolutionX) / 2;
+          }
+          else if (originalRatio < newRatio) {
+            var resizeFactor = size.width / resolutionX;
+            cropWidth = resolutionX;
+            cropHeight = size.height / resizeFactor;
+            cropY = (cropHeight - resolutionY) / 2;
+          }
+
+          this.options({imageMagick : true})
+            .autoOrient()
+            .resize(cropWidth, cropHeight)
+            .crop(resolutionX, resolutionY, cropX, cropY)
+            .stream(fileType, function (err, stdout, stderr) {
+              var r = stdout.pipe(res);
+              r.on('finish', function () {
+                // This is to close the result while a background job will continue to process
+                log('info', 'Finished sending a converted image');
+                res.end();
+              });
```

Comment:
If you pipe one stream into another then the target stream is already automatically closed when the source closes. Additionally, `stream.pipe` returns the *target* pipe, not the source pipe. So this is effectively `res.on('finish', () => res.end)`, which is a no-op.